include/verilated_funcs.h

```diff
@@ -2228,64 +2228,261 @@ static inline WDataOutP VL_SEL_WWII(int obits, int lbits, WDataOutP owp, WDataIn
+    VlQueue<T> ret;
+    VL_REVCOPY_Q(ret, from, lbits, elementBits);
+    return ret;
+}
+
+template <typename T>
+static inline void VL_COPY_Q(VlQueue<T>& q, const VlQueue<T>& from) {
+    q.renew(from.size());
+    for (size_t i = 0; i < from.size(); ++i) q.atWrite(i) = from.at(i);
+}
+
+// Helper function to extract a bit from a flat bit stream represented by a queue
+template <typename T>
+static inline bool VL_GET_QUEUE_BIT(const VlQueue<T>& queue, int elementBits,
+                                    size_t bitIndex) VL_PURE {
+    const size_t elementIndex = bitIndex / elementBits;
+    const size_t bitInElement = bitIndex % elementBits;
+    if (elementIndex >= queue.size()) return false;
+    return (queue.at(elementIndex) >> bitInElement) & 1;
+}
+
+// Helper function to set a bit in a flat bit stream represented by a queue
+template <typename T>
+static inline void VL_SET_QUEUE_BIT(VlQueue<T>& queue, int elementBits, size_t bitIndex,
+                                    bool value) VL_MT_SAFE {
+    const size_t elementIndex = bitIndex / elementBits;
+    const size_t bitInElement = bitIndex % elementBits;
+    if (elementIndex >= queue.size()) return;
+
+    T& element = queue.atWrite(elementIndex);
+    if (value) {
+        element |= (static_cast<T>(1) << bitInElement);
+    } else {
+        element &= ~(static_cast<T>(1) << bitInElement);
+    }
+}
+
+template <typename T>
+static inline void VL_REVCOPY_Q(VlQueue<T>& q, const VlQueue<T>& from, int lbits,
+                                int elementBits) {
+    // Flat bit reversal: treat the entire queue as a flat bit stream and reverse it in blocks of
+    // lbits
+
+    if (&q == &from) {
+        // In-place operation - need temporary copy
+        VlQueue<T> temp = from;
+        q.renew(temp.size());
+
+        const size_t totalBits = temp.size() * elementBits;
+        const size_t numBlocks = (totalBits + lbits - 1) / lbits;
+
+        // Clear the output queue
+        for (size_t i = 0; i < q.size(); ++i) { q.atWrite(i) = 0; }
+
+        // Reverse blocks
+        for (size_t block = 0; block < numBlocks; ++block) {
+            const size_t srcBlockStart = block * lbits;
+            const size_t dstBlockStart = (numBlocks - 1 - block) * lbits;
+
+            // Copy bits within the block
+            for (int bit = 0; bit < lbits; ++bit) {
+                const size_t srcBitIndex = srcBlockStart + bit;
+                const size_t dstBitIndex = dstBlockStart + bit;
+
+                if (srcBitIndex < totalBits && dstBitIndex < totalBits) {
+                    bool bitValue = VL_GET_QUEUE_BIT(temp, elementBits, srcBitIndex);
+                    VL_SET_QUEUE_BIT(q, elementBits, dstBitIndex, bitValue);
+                }
+            }
+        }
+    } else {
+        q.renew(from.size());
+
+        const size_t totalBits = from.size() * elementBits;
+        const size_t numBlocks = (totalBits + lbits - 1) / lbits;
```

Comment:
Duplicates code above in the `if` branch.  Use a reference or otherwise remove duplication?